lib/SILOptimizer/Mandatory/DataflowDiagnostics.cpp

```diff
@@ -117,35 +119,62 @@ static void diagnoseStaticReports(const SILInstruction *I,
 // SWIFT_ENABLE_TENSORFLOW
 /// \brief Emit a diagnostic for `poundAssert` builtins whose condition is
 /// false or whose condition cannot be evaluated.
-/// TODO(clattner / marcrasi): Implement this for real.
-static void diagnosePoundAssert(const SILInstruction *I, SILModule &M) {
+static void diagnosePoundAssert(const SILInstruction *I, SILModule &M,
+                                ConstExprEvaluator &constantEvaluator) {
   if (auto *builtinInst = dyn_cast<BuiltinInst>(I)) {
     if (builtinInst->getBuiltinKind() == BuiltinValueKind::PoundAssert) {
-      diagnose(M.getASTContext(), I->getLoc().getSourceLoc(),
-               diag::pound_assert_not_implemented);
+      SmallVector<SymbolicValue, 1> values;
+      constantEvaluator.computeConstantValues({builtinInst->getArguments()[0]},
+                                              values);
+      SymbolicValue value = values[0];
+      if (!value.isConstant()) {
+        // TODO(marcrasi / clattner): Emit more informative error.
+        diagnose(M.getASTContext(), I->getLoc().getSourceLoc(),
+                 diag::pound_assert_condition_not_constant);
+        return;
+      }
+      if (value.getTypeKind() != SymbolicValue::TKInteger) {
+        llvm_unreachable("sema prevents non-integer #assert condition");
+      }
+
+      APInt intValue = value.getIntegerValue();
+      if (!intValue.isSignedIntN(1)) {
+        llvm_unreachable("sema prevents non-int1 #assert condition");
+      }
+      if (intValue.isNullValue()) {
+        // TODO(marcrasi / clattner): Emit the actual #assert message.
+        diagnose(M.getASTContext(), I->getLoc().getSourceLoc(),
+                 diag::pound_assert_failure);
+        return;
+      }
     }
   }
 }
 
 namespace {
-class EmitDFDiagnostics : public SILFunctionTransform {
+class EmitDFDiagnostics : public SILModuleTransform {
```

Comment:
If I create one per function, it hangs while building the stdlib (or at least takes a really long time).

I changed it to create one per `pound_assert` instruction, which fixes that.

Alternatively, I could change it to make at most one per function -- only for functions that contain `pound_assert`. But I am guessing that we'll eventually move it out into a single place that can be shared by all passes in the module, so that would be throwaway work. Is that right?